### PR TITLE
Add support for Cargo private registries

### LINKED ISSUE #3478: Cargo private registry support
Cargo scanning doesn't seem to support private cargo registries. 

It'd be great to be able to add something like this;

```yaml
registries:
  cargo-example:
    type: cargo-registry
    url: https://dl.cloudsmith.io/basic/OWNER/REPOSITORY/cargo/index.git
    username: cargo-user
    password: ${{secrets.CARGO_GIT_PASSWORD}}
```

### PR DESCRIPTION
**Pull request by CodingAnarchy**
Working to resolve #3478 and enable the support for private Cargo registries. I built this off of #6868, which appeared to be abandoned a while ago in an incomplete state.

I've used the `bin/dry-run.rb` to validate this works for Cargo's sparse registries in the dry run scenario (using a private repository with Cloudsmith), and added a bit of clean up and unit testing.

### CODE REVIEW

**Comment by CodingAnarchy** [comment]
Currently reaching out to @cloudsmith-io. It appears that we are stymied with the Cargo download endpoint - `crates.io` returns metadata when the full version path is not specified to download the crate file, but private registries do not all follow this pattern. I'm trying to work with them to provide some API expectation, such as `/api/{crate}`, that could be what Dependabot expects for private registries to fetch the latest version.

At present this would _unbreak_ Dependabot failing to understand the private registry, but will not allow for updates for those registries(because it cannot resolve the latest version). I think this is still useful for a partial fix, as right now it is completely inoperable for those of us with a single dependency linked to a private registry, but obviously not a complete resolution.

**Comment by BartoszBlizniak** [comment]
Hey @CodingAnarchy 👋 Bart from Cloudsmith here, I have responded to your support query from our new portal for which you should have received an email. We apologise for the inconvenience and looking to help you out asap!

**Comment by CodingAnarchy** [comment]
@jeffwidman I saw you had commented on the previous PR I branched this from offering to answer questions and provide support. Is there anything that I can do to accelerate the review process here? This is blocking some teams at Shopify and we would love to get it sorted.

Please let me know if there is anything I need to update to improve this PR so it can be accepted.

**Comment by carogalvin** [comment]
@CodingAnarchy hey, I'm the PM for Dependabot here at GitHub. We're excited about this contribution from you! There are some steps we need to take on our internal repos for it to work once it's merged which we can't commit to doing right now 😓 

HOWEVER, we will be able to do this next quarter (April-ish) and would like to revisit merging this at that time. Would that work for you? I apologize for our internal processes slowing this down!

**Comment by CodingAnarchy** [comment]
@carogalvin I don't think we are on a tight timetable for it fully working, but I know that presently the introduction of the (non-functional) private registry breaks Cargo being able to update public dependencies as well. If there is a way that we could get that piece of it resolved sooner, we would really appreciate it, but I understand the need to get that slotted in the roadmap. Thanks!

**Comment by carogalvin** [comment]
@CodingAnarchy OK, are you able to submit a GitHub support request for the problem with private registries in the meantime? That's the best way to get it into our team's process and then hopefully we can solve that bug a little sooner then pick this up next quarter to fix properly.

**Comment by marosset** [comment]
> @CodingAnarchy OK, are you able to submit a GitHub support request for the problem with private registries in the meantime? That's the best way to get it into our team's process and then hopefully we can solve that bug a little sooner then pick this up next quarter to fix properly.

I'm also waiting for this work to land and can submit a GItHub support request for this too

**Comment by CodingAnarchy** [comment]
@carogalvin I've opened a support ticket with some of the details of the error that is blocking Dependabot for us. Please let me know if there is anything else you need from me to help get this sorted. Really appreciate the help!

File: cargo/lib/dependabot/cargo/file_fetcher.rb
Comment (Nishnha) [inline cargo/lib/dependabot/cargo/file_fetcher.rb:333]: Can you explain this bit? Can the config file be named `config` or `config.toml`?

File: cargo/lib/dependabot/cargo/file_parser.rb
Comment (Nishnha) [inline cargo/lib/dependabot/cargo/file_parser.rb:206]: Generally we want to avoid passing tokens in the Dependabot ruby Updater because they can show up in logs or 3rd party issue trackers.

Pub is the most recent ecosystem that we enabled private registries support, if you want to take a look at that implementation as an example, although it only supports git registries AFAIK https://github.com/dependabot/dependabot-core/blob/main/pub/lib/dependabot/pub/helpers.rb#L235

**Comment by Nishnha** [comment]
> but I know that presently the introduction of the (non-functional) private registry breaks Cargo being able to update public dependencies as well. If there is a way that we could get that piece of it resolved sooner (I'm not sure if merging this would do it without fixing your internal code as well),

Would it help if we split this PR into 2 parts, one for unblocking updates that only use the public registry, and another for actually adding private registry support (which will require us to fix our internal code)? If we do that, adding support for git registries should be relatively easy.

Otherwise we will have to wait until the team can fund fixing the internal code to support cargo URLs which still looks like April.

File: cargo/lib/dependabot/cargo/file_fetcher.rb
Comment (CodingAnarchy) [reply cargo/lib/dependabot/cargo/file_fetcher.rb:333]: Cargo will read the `.cargo/config.toml` file, and fallback to the `.cargo/config` file as the `toml` file extension support was added in v1.39. This order here is backwards, so I am pushing up a change to fix that, but otherwise will be necessary to ensure that we pull in the file that Cargo expects even on older repositories.

File: cargo/lib/dependabot/cargo/file_parser.rb
Comment (CodingAnarchy) [reply cargo/lib/dependabot/cargo/file_parser.rb:206]: I'm open to suggestions to avoid that here, but I'm not sure how else to fetch the metadata for the crates without reaching out to the index API that Cargo uses internally. I'm using the pattern in your reference link to have `cargo` do the actual updates of the dependencies, but is there another way you would recommend to reach out the API and get a JSON response that avoids this use of the token?

**Comment by CodingAnarchy** [comment]
> Would it help if we split this PR into 2 parts, one for unblocking updates that only use the public registry, and another for actually adding private registry support (which will require us to fix our internal code)? 

I'm not sure that it can be split as updated here. The reason this fixes the public registry case is that `cargo` is now aware from the config that the private registry has a specified index URL. We _may_ be able to just do the parts that move the `.cargo/config.toml` file into the part doing the update and unblock updates (assuming I am not missing something about how Dependabot or Cargo is managing that part), though I am not sure that I will be able to do that quickly either myself. I'm about to go out on parental leave until April, so I have no idea how much time I will have to split the PRs.

File: cargo/lib/dependabot/cargo/file_fetcher.rb
Comment (jeffwidman) [reply cargo/lib/dependabot/cargo/file_fetcher.rb]: To add context here, Dependabot requires the file to be in `toml`-parsable format: 
* https://github.com/dependabot/dependabot-core/pull/7120

Are the `.cargo/config` files still in `toml` format but just lacking the file extension? Or are they a different format altogether?

Comment (CodingAnarchy) [reply cargo/lib/dependabot/cargo/file_fetcher.rb]: Same file format, it is just that Cargo did not originally require the extension.

**Comment by pavera** [comment]
@CodingAnarchy I've opened a PR to at least start to address the issue where defining a private registry breaks dependabot: https://github.com/dependabot/dependabot-core/pull/9109

I'm not sure if that will fully unblock you, but I think it will at least get past the initial cargo error where it can't find the definition of the private registry.

File: cargo/lib/dependabot/cargo/file_parser.rb
Comment (mctofu) [reply cargo/lib/dependabot/cargo/file_parser.rb:206]: I've added support for sparse registry in our proxy: https://github.com/dependabot/dependabot-core/issues/3478#issuecomment-1973940575

We'd need to test this out using the CLI rather than the dry-run script as that will be a more accurate representation of how this will run in the GitHub hosted version. The differences from the dry-run script will be:
- The credentials passed in here will only have the type & url fields set but not token. (registry name could be provided too if necessary)
- The addition of the token to the request will happen in a separate process by the proxy

It would be OK to remove the token auth support from this code here unless you plan to run this in a standalone mode without the proxy.

**Comment by RobJellinghaus** [comment]
Hello @CodingAnarchy and @pavera, I work for Microsoft on internal Rust support, and I have been testing this PR against our internal private registries. I forked it and created my own branch based on this one, and merged it up to date with dependabot-core main.

After some bug fixes, I'm happy to say it's mostly working in my current testing, including using `dependabot-cli` and the Dependabot credential proxy. So @mctofu was correct [when he said that this was working](https://github.com/dependabot/dependabot-core/issues/3478#issuecomment-1973940575). There is still an issue with updating a package that exists only in a private registry, but it may be a bug in Azure Artifacts; I'm prosecuting that internally.

My question for @CodingAnarchy is: are you still driving this PR? I see from your GitHub activity that you've been back online for all of April, but I haven't seen any activity on this branch. I don't want to create a new PR unless you are stepping back from this one, since you did do most of the work and my fixes are relatively small. Please advise? If I don't hear back from you within a week (by May 7th), I will create a new PR.

For reference, my fork and branch are here: https://github.com/RobJellinghaus/dependabot-core/tree/user/rjelling/cargo-private-registries

**Comment by RobJellinghaus** [comment]
(also tagging @jakecoffman, @abdulapopoola, and @brbayes-msft as FYI)

**Comment by CodingAnarchy** [comment]
Hey @RobJellinghaus! I just got reminded of this the other day, and was about to resolve this against the changes in main. Instead, I've just merged up to your forked branch, which made it quite easy to resolve. I was able to replicate the results of your testing locally as well, so my understanding now is that the @pavera and the rest of the Dependabot team should be good to merge this.

@pavera @mctofu If there is anything else that I can do to enable this support, please let me know!

**Comment by honeyankit** [comment]
@CodingAnarchy I am going to deploy this PR. There is still some work left from our side on the API side, which I have started already. Thank you everyone who has worked on this PR.

**Review by honeyankit — APPROVED** [review]

**Comment by honeyankit** [comment]
@RobJellinghaus Could you share the `job.yml` file which you used for testing which passes with the cargo private registry based on the [comment](https://github.com/dependabot/dependabot-core/pull/8719#issuecomment-2087675841)?

**Comment by RobJellinghaus** [comment]
Hello @honeyankit, let me guess, the parse error you are getting is similar to this one?
```
updater | 2024/05/06 20:46:12 ERROR Error processing nuget (JSON::ParserError)
updater | 2024/05/06 20:46:12 ERROR unexpected token at '{"name":"nuget","vers":"0.2.1","deps":[{"name":"anyhow","req":"^1.0.40","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"bytes","req":"^1.4.0","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"home","req":"^0.5.3","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"indicatif","req":"^0.17","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"junction","req":"^1.0.0","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"reqwest","req":"^0.11","features":["json","blocking"],"optional":false,"default_features":false,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"reqwest-middleware","req":"^0.2.0","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"reqwest-retry","req":"^0.4.0","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"semver","req":"^1","features":["serde","serde"],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"serde","req":"^1","features":["derive","derive"],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"serde_json","req":"^1","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"service_error","req":"^0.1.1","features":["reqwest"],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":null,"package":null},{"name":"thiserror","req":"^1","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"tokio","req":"^1.26.0","features":["macros","sync","time","fs","rt-multi-thread"],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"tracing","req":"^0.1.37","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"zip","req":"^0.6.2","features":["deflate"],"optional":false,"default_features":false,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"mockito","req":"^1.1.0","features":[],"optional":false,"default_features":true,"target":null,"kind":"dev","registry":"https://github.com/rust-lang/crates.io-index","package":null}],"cksum":"c5a9dfd7380a70e46a876ed67ae11d8e67bc92bcc5e626b0a05f14f3d58d0cac","features":{},"yanked":false,"links":null,"v":2}
updater | {"name":"nuget","vers":"0.1.0","deps":[{"name":"anyhow","req":"^1.0.40","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"bytes","req":"^1.4.0","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"home","req":"^0.5.3","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"indicatif","req":"^0.17","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"junction","req":"^1.0.0","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"reqwest","req":"^0.11","features":["json","blocking"],"optional":false,"default_features":false,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"reqwest-middleware","req":"^0.2.0","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"reqwest-retry","req":"^0.3.0","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"semver","req":"^1","features":["serde","serde"],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"serde","req":"^1","features":["derive","derive"],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"serde_json","req":"^1","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"thiserror","req":"^1","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"tokio","req":"^1.26.0","features":["macros","sync","time","fs"],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"tracing","req":"^0.1.37","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"zip","req":"^0.6.2","features":["deflate"],"optional":false,"default_features":false,"target":null,"kind":"normal","registry":"https://github.com/rust-lang/crates.io-index","package":null},{"name":"mockito","req":"^1.1.0","features":[],"optional":false,"default_features":true,"target":null,"kind":"dev","registry":"https://github.com/rust-lang/crates.io-index","package":null}],"cksum":"e678fc543007ade32539623426e912dd9544c82fde2627c9b3d4f31cd43f0994","features":{},"yanked":false,"links":null,"v":2}
updater | '
updater | 2024/05/06 20:46:12 ERROR /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/gems/json-2.6.3/lib/json/common.rb:216:in `parse'
updater | 2024/05/06 20:46:12 ERROR /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/gems/json-2.6.3/lib/json/common.rb:216:in `parse'
```

I am sorry that I did not alert you to this problem sooner, I was actually aware of this a couple of weeks ago. But because the root issue involves some more work in dependabot-core, I wanted to sync with the GitHub Dependabot team before bringing this up. However, they were in an offsite last week and the next sync opportunity is early next week, so I was letting this slide until then. That was a bad call and I apologize.

The parse error is because dependabot-core is trying to fetch the latest version of a package from a private registry, and the response is actually not JSON. The Cargo registry protocol is specified to return a _sequence_ of JSON objects (e.g. the Cargo index file for a crate is a text file with one JSON object per line): https://doc.rust-lang.org/cargo/reference/registry-index.html#index-files 

But dependabot-core does not know this, and thinks the whole response is invalid.

The root problem here is that it turns out dependabot-core has always done the wrong thing when it comes to looking up the latest version of a package. Currently dependabot-core actually hits the public crates.io web site -- not the crates.io sparse registry! -- to determine the latest version of a crate. This is wrong, for three reasons:

- The public crates.io website's JSON format is not specified and subject to change, so trying to scrape it is not supported.
- Private registries don't support the crates.io website endpoint, and so dependabot-core can't actually handle their responses.
- From a Microsoft internal perspective, we filter all public crates through a private registry, and we may quarantine or yank crates that the public crates.io service has published. So we _can't_ go to the public crates.io as the source of truth for public crates.

This all means that we actually need a new issue, "Dependabot-core Cargo implementation must support Cargo index file format" -- this was not on your radar when you implemented this, and as I said I did not do a good job of promptly communicating the issue two weeks ago. I will be syncing with the GitHub team early next week and will make the issue immediately thereafter, if you don't beat me to it. Unfortunately I am personally heads down in urgent internal security work for the time being (duration TBD), otherwise I would be digging in here already.

The job.yml file is not the problem here, but for completeness, mine looked like this:
```
job:
  package-manager: cargo
  allowed-updates:
  - dependency-type: direct
    update-type: all
  source:
    provider: azure
    repo: mscodehub/Rust/_git/rust.crate-knowledge
    directory: "/."
    branch: user/rjelling/1.75

credentials:
- type: cargo_registry
  url: https://pkgs.dev.azure.com/mscodehub/Rust/_packaging/Rust_PublicPackages/Cargo/index
  cargo_registry: Rust_PublicPackages
  token: Basic [base64-encoded PAT]

- type: cargo_registry
  url: https://pkgs.dev.azure.com/mscodehub/Rust/_packaging/Rust_CrateReview/Cargo/index
  cargo_registry: Rust_CrateReview
  token: Basic [base64-encoded PAT]
```

**Comment by honeyankit** [comment]
@RobJellinghaus : No worries. Gald we catched it earlier

My error is coming pretty early in the paring the cargo file_parser. I will have to look into what is the issue. Based on your job.yml file I changed mine to include as cloudsmith provide Basic auth but same issue. 

```bash
credentials:
  - type: cargo_registry
  - registry: honeyankit-test
    url: "https://cargo.cloudsmith.io/honeyankit/test/"
    token: "Basic <token>"
```

```bash
024/05/17 21:28:00 INFO Starting job processing
2024/05/17 21:28:00 ERROR undefined method `name' for nil
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:295:in `parsed_file'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:237:in `cargo_config_from_file'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:228:in `cargo_config_field'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:176:in `registry_source_details'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:169:in `source_from_declaration'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:271:in `git_req?'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:258:in `block in version_from_lockfile'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:257:in `select'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:257:in `version_from_lockfile'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:72:in `block (3 levels) in manifest_dependencies'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:70:in `each'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:70:in `block (2 levels) in manifest_dependencies'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:69:in `each'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:69:in `block in manifest_dependencies'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:68:in `each'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:68:in `manifest_dependencies'
2024/05/17 21:28:00 ERROR /home/dependabot/cargo/lib/dependabot/cargo/file_parser.rb:30:in `parse'
2024/05/17 21:28:00 ERROR /home/dependabot/dependabot-updater/lib/dependabot/dependency_snapshot.rb:197:in `parse_files!'
20
```

**Comment by honeyankit** [comment]
@RobJellinghaus: Could you also share the cargo.toml and cargo.json file if you have it? I want to have look to both them.

**Comment by RobJellinghaus** [comment]
Also tagging @CodingAnarchy as my apologies also should go to him, since he landed the PR partly based on my claim that it worked. Which it did, as far as it went... but this problem was exposed.

**Comment by RobJellinghaus** [comment]
> @RobJellinghaus: Could you also share the cargo.toml and cargo.lock file if you have it? I want to have look to both them.

I'm sorry, that I can't do -- this is not open source, it's a Microsoft repository not a public GitHub one. 

If you have a Microsoft / Azure DevOps account, you can fetch it directly, please email me (at my Microsoft account, rjelling) for the details there.

**Comment by honeyankit** [comment]
@RobJellinghaus: My issue is that it is looking to get the `index_url = cargo_config_field("registries.#{registry_name}.index")` from the `.cargo/config.toml` which I have not created in my test repo. I do not think we need the `.cargo/config.toml` to create the index.



```bash
      def cargo_config
        @cargo_config ||= get_original_file(".cargo/config.toml")
      end
```

**Comment by honeyankit** [comment]
I think I have fixed it with my initial changes:

```bash
Fetched metadata for hello-world from sparse+https://cargo.cloudsmith.io/honeyankit/test/ successfully
{"name": "hello-world", "vers": "1.0.0", "deps": [], "cksum": "b2c263921f1114820f4acc6b542d72bbc859ce7023c5b235346b157074dcccc7", "features": {}, "yanked": false, "links": null}
{"name": "hello-world", "vers": "1.0.1", "deps": [], "cksum": "8a55b58def1ecc7aa8590c7078f379ec9a85328363ffb81d4354314b132b95c4", "features": {}, "yanked": false, "links": null}
2024/05/17 22:53:12 INFO Latest version is 1.0.1
2024/05/17 22:53:12 INFO No update needed for hello-world 1.0.1
2024/05/17 22:53:12 INFO Finished job processing
```

**Comment by honeyankit** [comment]
Issue in this [comment](https://github.com/dependabot/dependabot-core/pull/8719#issuecomment-2118373910) is not fixed: https://github.com/dependabot/dependabot-core/pull/9783/files#r1607372525

File: cargo/lib/dependabot/cargo/file_parser.rb
Comment (jeffwidman) [reply cargo/lib/dependabot/cargo/file_parser.rb:206]: Follow up PR's addressing ☝️ of changing how token auth was handled:
* https://github.com/dependabot/dependabot-core/pull/14030 - superseded by 👇 
* https://github.com/dependabot/dependabot-core/pull/14340

**Review by leuasseurfarrelds247-arch — COMMENTED** [review]
4bafd3b8b895718f23665e891efcd696bcacc387

**Pull request merged**
